core/model/asynctask.py

```diff
@@ -0,0 +1,115 @@
+# Async
+import datetime
+import json
+import uuid
+from enum import Enum
+from typing import Any
+
+from pydantic.dataclasses import dataclass
+from sqlalchemy import Column, DateTime, String
+from sqlalchemy.dialects.postgresql import JSONB, UUID
+from sqlalchemy.ext.mutable import MutableDict
+from sqlalchemy.types import Enum as SqlAlchemyEnum
+
+from core.model import Base, create
+from core.util.datetime_helpers import utc_now
+
+
+class AsyncTaskStatus(str, Enum):
+    READY = "READY"
+    PROCESSING = "PROCESSING"
+    SUCCESS = "SUCCESS"
+    FAILURE = "FAILURE"
+
+
+class AsyncTaskType(Enum):
+    INVENTORY_REPORT = "INVENTORY_REPORT"
+
+
+class AsyncTask(Base):
+    """An asynchronous task."""
+
+    __tablename__ = "asynctasks"
+    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
+    created = Column(
+        DateTime(timezone=True), index=True, nullable=False, default=utc_now
+    )
+    task_type = Column(SqlAlchemyEnum(AsyncTaskType), index=True, nullable=False)
+    status = Column(SqlAlchemyEnum(AsyncTaskStatus), index=True, nullable=False)
+    processing_start_time = Column(DateTime(timezone=True), nullable=True)
+    processing_end_time = Column(DateTime(timezone=True), nullable=True)
+    status_details = Column(String, nullable=True)
+    data: dict[str, Any] = Column(MutableDict.as_mutable(JSONB), default={})
+
+    def complete(self):
+        if self.status != AsyncTaskStatus.PROCESSING:
+            raise Exception(
```

Comment:
It looks like we're catching `ProblemDetailException`s in the controller, so perhaps we should be throwing a `ProblemDetailException` here. Even if we don't need them to be `ProblemDetailException` it would be nice to have these live in the `BasePalaceException` hierarchy.